File: Chapter2/genby2gram.rb
#! /usr/local/bin/ruby -Ku
# -*- coding: utf-8 -*-

def read2gram(db2gram)
  i = 0
  open("2gram.txt") {|file|
    while l = file.gets
      db2gram[i] = l.chomp
      i += 1
    end

  }

  return i 
end



def findch(startch, db2gram, n)
  no = 0
  for i in 0..n-1
    if startch==db2gram[i][0] then

      no += 1
    end
  end
  return no
end

def setrnd(num)
  return rand(num)
end



def setrndstr(startch, db2gram, n)
  point = setrnd(n)
  for i in 0..n-1
    if i==point then
      startch[0] = db2gram[i][1]
    end
  end
end



def setnext(startch, db2gram, n, num)
  point = setrnd(num)+1
  no = 0
  for i in 0..n
    if startch==db2gram[i][0] then
      no += 1
    end
    if no == point then
      startch[0] = db2gram[i][1]
      break
    end
  end
end



def generates(inputString, db2gram, n)
  startch = inputString.dup
  print startch[0]
  begin
    num = findch(startch, db2gram, n)
    if num != 0 then
      setnext(startch, db2gram, n, num)
    else
      setrndstr(startch, db2gram, n)
    end
   print startch[0]
 end while startch[0]!='．' && startch[0]!='，'
  puts ""
end


#main function
MAXNO = 100000
MAXLINE = 256

db2gram = Array.new(MAXNO, '\0')

num = read2gram(db2gram)

puts "開始文字を入力してください。"
inputString = STDIN.gets 
inputString = inputString.chomp

for i in 0..10
  generates(inputString, db2gram, num)
end

File: sample/ch2/genby2gram.c
//	genby2gram.c
//	2-gramの連鎖により文を作成する
//	2-gramが格納されたファイル2gram.txtを用います
//	開始文字を指定すると文を生成します

#include<stdio.h>
#include<string.h>
#include<stdlib.h>

#define MAXNO 100000 //2-gramの最大数
#define MAXLINE 256 //1行の最大バイト数
#define FILENAME "2gram.txt" //2-gramが格納されたファイル

/*2-gramファイルの読み込み*/
int read2gram(char db2gram[MAXNO][6])
{
 FILE *fp;
 char line[MAXLINE] ;
 int i=0 ;

 if((fp=fopen(FILENAME,"r"))==NULL){
  fprintf(stderr,"エラー　ファイル2gram.txtがありません\n");
  exit(1) ;
 }

 while(fgets(line,MAXLINE,fp)!=NULL){
  strncpy(db2gram[i],line,4) ;//全角2文字をコピー
  db2gram[i][5]='\0' ;
  ++i ;
 }
 return i ;//2-gramの個数を返す
}
/*開始文字が何回含まれるか数える*/
int findch(char *startch,char db2gram[MAXNO][6],int n) 
{
 int i ;
 int no=0 ;
 for(i=0;i<n;++i){
  if(strncmp(startch,db2gram[i],2)==0) ++ no ;
 }
 return no ;
}
/*num未満の乱数をセット*/
int setrnd(int num)
{
 int rndno ;
 while((rndno=(double)rand()/RAND_MAX*num)==num) ;
 return rndno ;

}

/*次の文字をn-gramによりランダムにセット*/
void setrndstr(char *startch,char db2gram[MAXNO][6],int n)
{
 int i ;
 int point ;
 
 point=setrnd(n) ;//n未満の乱数をセット
 for(i=0;i<n;++i){
  if(i==point){
   startch[0]=db2gram[i][2] ;
   startch[1]=db2gram[i][3] ;
   startch[2]='\0' ;
   break ;
  }
 }  
}

/*次の文字をn-gramによりセット*/
void setnext(char *startch,char db2gram[MAXNO][6],int n,int num) 
{
 int i ;
 int no=0 ;
 int point ;
 
 point=setrnd(num) ;//num未満の乱数をセット
 for(i=0;i<n;++i){
  if(strncmp(startch,db2gram[i],2)==0) ++ no ;
  if(no==point){
   startch[0]=db2gram[i][2] ;
   startch[1]=db2gram[i][3] ;
   startch[2]='\0' ;
   break ;
  }
 }  
}


/*文の生成*/
void generates(char *startch,char db2gram[MAXNO][6],int n)
{
 int i,num ;
 
  /*開始文字を出力する*/
  putchar(startch[0]) ; putchar(startch[1]) ;
  /*句点が出るまで繰り返し*/
 do{
  /*開始文字が何回含まれるか数える*/
  num=findch(startch,db2gram,n) ;
  /*その中からランダムに文字列を選ぶ*/
  if(num!=0)
   setnext(startch,db2gram,n,num) ;
  else
   setrndstr(startch,db2gram,n) ;
  /*文字を出力する*/
  putchar(startch[0]) ; putchar(startch[1]) ;
 }while((strncmp(startch,"．",2)!=0)&&(strncmp(startch,"。",2)!=0)) ;
 printf("\n") ;
}

 



int main()
{
 char db2gram[MAXNO][6] ;//2gramのデータベース
 int n ;//2-gramの個数
 char startch[MAXLINE];//開始文字
 int i ;
 char workch[MAXLINE];//作業用の文字

 
 /*乱数の初期化*/
 srand(65535) ;
 /*2-gramファイルの読み込み*/
 n=read2gram(db2gram) ;


 /*開始文字の決定*/
 printf("開始文字を入力してください\n") ;
 fgets(startch,MAXLINE,stdin) ;

 /*10回の文の生成*/
 for(i=0;i<10;++i){
  strcpy(workch,startch) ;
  generates(workch,db2gram,n) ;
 }
 return 0 ;
}



